### ISSUE TITLE
Feature: ability to move view! macro to a separate file

### ISSUE THREAD

**Issue by azymohliad** [Relm4/Relm4#246]
The `view! { ... }` macro can be gigantic sometimes, and it's not always reasonable/easy to split things into smaller components. Having `view! { ... }` in a separate file (from the rest of the component implementation) can make the code much easier to navigate and maintain in such cases. Additionally, it would provide even cleaner separation between declarative and imperative parts.

**Comment by AaronErhardt** [comment]
Sounds like a good idea, but first we need to know how this would look like.

My first though was to use a new `view_file!()` macro:

```rust

#[relm4::component]
impl SimpleComponent for AppModel {
    type Init = u8;
    type Input = AppMsg;
    type Output = ();
    type Widgets = AppWidgets;

    view_file!("path/to/view.rs")

   ...
}
```

And then in `view.rs`:

```rust
view! {
    // regular view macro, as if it was created inside of `#[relm4::component]`
}
```

Maybe we could use rust modules as path specifier instead. What do you think?

**Comment by azymohliad** [comment]
Hm.. If it's identified by module path, should it actually be a proper Rust module? I guess yes, otherwise the path is invalid from the compiler PoV. But then, if it's a different module, all component-level types accessed from view macro would need to be properly qualified there (e.g with `super::` if view module is child to component module). At the same time, view macro can access local variables and arguments of component methods, which breaks module's access boundaries.

I guess it'd be less confusing to just treat it as `include!()` file, and so identify by file path. It would also allow more freedom experimenting with a file structure, e.g. grouping all view files together in one folder (not sure if it's a good idea tho :P).

But I might be wrong at any stage of my chain of thoughts here. I never wrote complicated macros, and I don't have expertise to think about it on the level you can. So if module path won't cause those problems I imagined, then I'm completely happy with it too.

**Comment by AaronErhardt** [comment]
It probably makes sense to experiment a bit with this. I can try out how easy it is to add in general and then we'll see how things like IDE integration (e.g. rust-analyzer) will work out.

**Comment by AaronErhardt** [comment]
Do you want to check out #247 ? You could try out the `external-view` branch already if you want, but error reporting is horrible. Maybe we need to use Rust modules after all so that the compiler can figure out what to do with the error spans...

**Comment by azymohliad** [comment]
Wow, so fast! Unfortunately I won't be able to try it out properly until Saturday (UTC+3), I might quickly check tomorrow but not sure.

**Comment by azymohliad** [comment]
I finally tried the branch, everything works with my project. But yeah, error reporting and integration with rust analyzer are very good points in favor of modules.

**Comment by AaronErhardt** [comment]
Now that [widget templates](https://relm4.org/book/next/widget_templates.html) exist, I wonder whether those are the better option. They don't allow you to move all of the `view` macro into a different file, but certainly most of it and do so without any macro magic.

@azymohliad what do you think? Should we continue this effort or are widget templates good enough?

**Comment by azymohliad** [comment]
I'm happy with templates. Definitely they can help to reduce the boilerplate. I would guess that view macro in a separate file might be useful for some cases anyway. But as we discussed above, there are some issues, so maybe there's no point indeed.

I personally don't really need this feature that much anymore, because I've refactored my mess and actually separated reasonable components. I'm happy to close the issue, to be clear.

Sorry for a long silence, I've been either very busy or tired lately to look into templates properly.
